### ISSUE TITLE
PhET Menu - focus inconsistent when exiting the Menu with Tab and Shift Tab

### ISSUE THREAD

**Issue by terracoda** [phetsims/john-travoltage#213]
**Tab key and Shift Tab** 
- focus highlights work fine to go back and forth through the menu items. However,...
  - I don't hear item count when Tabbing (I do when using the Arrow keys). This could be a difference between menu navigation and Tab navigation.
  - when I hit Shift Tab on the first item, the pop-up closes (as it perhaps should if using the Tab key, but not if using the Arrow keys) and keyboard focus goes to the PhET Menu button.
- when I hit ESC the menu closes and focus goes to the PhET Menu button.
 - when I hit the Tab key on the bottom menu item, however, the pop-up closes, and the keyboard focus goes to the Keyboard button. Is that intentional? 

Should it not go to the PhET Menu Button in this case as well?

**Comment by jessegreenberg** [comment]
> when I hit the Tab key on the bottom menu item, however, the pop-up closes, and the keyboard focus goes to the Keyboard button. Is that intentional?

The intention was to focus the next focusable element in the document once the menu closes, like https://staff.washington.edu/tft/tests/menus/simplyaccessible/index.html

That menu item is the last thing in the document, so I expected focus to go back to the address bar - this is what it does in Chrome.  In Firefox and Safari, it looks like it is going to John Travoltage's foot. I haven't been able to get it to go back to the Keyboard button.

> Should it not go to the PhET Menu Button in this case as well?

That would definitely be the simplest solution, we can do that. I would like to know why Firefox and Safari don't do that though.

**Comment by jessegreenberg** [comment]
Hmm, not sure what the difference is between the browsers - could be an order of events for keydown listeners.

Anyway, we have two options:

- Focus back on the PhET Button when we tab out of the menu
- Focus the browser/address bar when we tab out of the menu

I have tested both, and they work consistently in browsers. I slightly prefer the second option, the user can "Shift+Tab" back to the last menu item if they leave the menu. But I am not sure what is most conventional.

@terracoda @emily-phet @jhung @jobara do you have a preference?

**Comment by terracoda** [comment]
@jessegreenberg, the convention is to move focus to the next focusable item (in this case the browser bar because the menu is the last item), if we follow this design pattern.

For a single menu button, I personally do not like this design pattern. I see how this design pattern could be useful for menus that have a few top-level items which have short submenus. In Safari with VoiceOver, the Left and Right Arrow keys do navigate the top level items as they do without a screen reader on.

For a single menu button, I prefer the W3C ARIA Practices design pattern, see [Action Menu widget](https://www.w3.org/TR/wai-aria-practices-1.1/examples/menu-button/menu-button-1/menu-button-1.html).

The W3C pattern allows users to cycle through items in the menu list without exiting it. They have to intentionally exits. 

In this pattern, Space and Enter keys open the Menu, Arrow keys are used to cycle through the menu items (they wrap from top to bottom and bottom to top), Tab key exits the pop-up menu and goes to next focusable item, Shift Tab exits the pop-up menu and goes to the previous focusable item, ESC key closes the pop-up menu and focus goes to the menu button.

The W3C example, however, still has issues in Safari with VoiceOver.

I have two suggests:
- quick option: Change the behaviour of the current implementation so that Tabbing on the last Menu item goes to the next focusble item (browser bar).
- longer option: implement the W3C pattern. 

We are working on this pattern in https://github.com/phetsims/a11y-research/issues/25

 Open to other's suggestions.

**Comment by jessegreenberg** [comment]
@terracoda, sorry I don't understand - what is the difference?

> the convention is to move focus to the next focusable item

> Tab key exits the pop-up menu and goes to next focusable item, Shift Tab exits the pop-up menu and goes to the previous focusable item

Aren't these the same pattern?

**Comment by terracoda** [comment]
@jessegreenberg, sorry for not being clear :-) I made some edits to the original post. They are in bold.

**Comment by terracoda** [comment]
Oops posted prematurely... The W3C convention is that the Tab key is **not** used to cycle through submenu items. 

Let's backup a little. We have two issues in our case, and I think this is what is causing our over all design problem with this bottom bar.

1. If the parent bottom bar is treated as a menu, then the PhET Menu button is a top-level item and the pop-up menu is a submenu.
    - unfortunately, in our case, we **only have two top-level items**, and only one of them has a submenu. The design pattern from Washington (which does not follow the W3C pattern) is not so beneficial with so few top-level items.

2. If we just think about the bottom bar as an important landmark with a couple of links in it, then the PhET Menu button fits the ARIA Practices _menubutton_ design pattern. It is the _menubutton_ design pattern that I was referring to as the longer option.
    - I think the _menubutton_ design pattern is the better design pattern for us, based on the research I did on Wednesday and Thursday last week in issue phetsims/a11y-research#25.

In either case, If we stick with using the Tab key to cycle through the items in the popped-up menu (i.e., not use ARIA Practices _menubutton_ pattern), then the Tab key does not close the popped-up menu. In that case, going backwards (Shift Tab) on the first item and forward (Tab key) on the last item should exit the menu and go to the previous focusable item (Keyboard Shortcuts button) and next focusable item (browser bar). ESC key on any item closes the popped-up menu and returns focus to the last focused item, the PhET Menu button.

@jessegreenberg, is that more clear?

**Comment by emily-phet** [comment]
@terracoda For clarification, are you proposing that we use option 2 from your comment?

**Comment by jessegreenberg** [comment]
Yes, understood @terracoda, thanks.

It would not be difficult at all to go with the W3C pattern (tab exits and closes the dialog instead of focusing the next item in the list). Should we do that?

**Comment by jessegreenberg** [comment]
The above commit makes the menu behave like the W3C example provided by @terracoda.  I tested with Chrome, Firefox, IE and Edge.

**Comment by terracoda** [comment]
@jessegreenberg and @emily-phet, sorry for the delay. Glad doing option 2 is not a lot of work. Based on my understanding thus far, I think it is the better option.

I spelled things out more clearly in https://github.com/phetsims/a11y-research/issues/25#issuecomment-292765290

Let me know when a new link is available for testing.

**Comment by jessegreenberg** [comment]
@terracoda is the keyboard navigation correct in this version?
 http://www.colorado.edu/physics/phet/dev/html/john-travoltage/1.3.0-dev.29/john-travoltage_en.html

**Comment by terracoda** [comment]
@jessegreenberg, here's a summary
**What's working:**
- Up and down Arrow keys are working properly, cycling through the menu items.
- Left Arrow key does nothing as it should. Keyboard focus stays on menu item. This is correct behaviour. VO reads out a long number, but that seems to be a VO idiosyncrasy.
- The semantics of a menu and menu item are read out, things like "You are on a menu item."
- Esc key works, and focus goes to PhET Menu as it should.
- Dialogs open and close properly, but on closing focus is lost (see item 2 below).
- All the items in the menu are currently operating like buttons. Enter and Space key are operable on all items. This is helpful since the semantics are not being read out. Enter and Space will be first guesses.

**Remaining Issues in VO and Safari:**
1. Pressing the Right Arrow should do nothing, but instead, visual focus is lost and the focus seems to go to the browser's back button. The PhET Menu stays open when focus is lost, and I have to Tab through the browser bar and back into the sim. When I get to the sim, the PhET pop-up Menu closes and the pink focus highlight is restored on the Leg swing.
  - Note pressing the Right Arrow and then Shift Tab, restores focus on the menu item that just lot focus. Pressing the Right Arrow, should do nothing.
  - Note the Right Arrow behaviour is the same (loses focus) when VO is off.
2. When the pop-up dialogs close, focus is lost. The virtual cursor seems to go to the H1. I think focus should go to the PhET Menu button when a dialog is closes (the last item to have had focus).
3. VO just craps out from time to time and focus is lost. Using the Tab key or Shift Tab gets me back into the sim.

@jessegreenberg, I may not have found everything, but if you can fix item item 1 (make the Right Arrow do nothing), and item 2 (ensure focus is returned to the PhET Menu button upon closing a dialog, it's good!

**Comment by jessegreenberg** [comment]
In #213 @terracoda also said 

> Also, any chance the keyboard focus highlight can also include the mouse focus style? Not a show stopper, that might be a longer term general issue we need to work on.

**Comment by terracoda** [comment]
@jessegreenberg, thanks for bringing #212 and #213 together.

**Comment by jessegreenberg** [comment]
> Pressing the Right Arrow should do nothing, but instead, visual focus is lost and the focus seems to go to the browser's back button

I can only get this to happen if VO is on or if I just turned VO off without resetting Safari.  It seems that Safari needs to be reset after VO is off to work normally.  We can try preventing default event on arrow keys, otherwise this is just movement of the virtual cursor.

**Comment by jessegreenberg** [comment]
Arrow key issue is fixed with the above commit.

**Comment by jessegreenberg** [comment]
> When the pop-up dialogs close, focus is lost. The virtual cursor seems to go to the H1. I think focus should go to the PhET Menu button when a dialog is closes (the last item to have had focus).

Thanks @terracoda, it is trying to set focus on the PhET menu item since that was the last thing focused, but it no longer exists in the document.  We can add an option that specifies where focus should go when a dialog is closed.

**Comment by emily-phet** [comment]
@jessegreenberg Regarding https://github.com/phetsims/john-travoltage/issues/213#issuecomment-297566892 I think resetting Safari before turning on VO fixed the issue I was running into where focus would move on its own while using the arm slider.

**Comment by terracoda** [comment]
@jessegreenberg, I tested the PhET Menu in version 31, and it seems to be working a little differently than before. Here is what I found:

**PhET Menu with keyboard and Voice Over off**
- Pressing Left Arrow does nothing on each menu item! Perfect!!!
- Pressing Right Arrow loses visual focus without closing PhET Menu. Pressing Shift Tab brings keyboard back, but always to the About menu item. I can navigate the items when focus comes back.
- Dialogs work, but focus is lost on closing the dialog, same as before.

**With Voice Over On**
- Pressing Left arrow key on menu items loses focus on the open menu.
  - Pressing Shift Tab goes to browser bar, so need to Tab through or Shift Tab through to get back to Menu button.
- Pressing Right arrow loses focus on all items. 
  - Pressing Shift Tab brings focus back to the About item.
  - Sometimes I cannot navigate the items with Up and Down arrow after focus returns to the About item. This seems to be random.
- When VO Konks out (randomly), I can't use the Up and Down Arrow keys to navigate the menu items. Usually a page reload fixes this issue. 
- I am hearing the semantics of a menu item sometimes, but never the inner semantics. I think this is the same as before.

**Comment by jessegreenberg** [comment]
@emily-phet good to hear, thanks!

@terracoda thanks, yes that is expected, version 31 doesn't include any fixes for this (specifically https://github.com/phetsims/joist/commit/d2bdbf79ca3c4f28b8e438efe34095ec4a053969).

**Comment by terracoda** [comment]
@jessegreenberg re https://github.com/phetsims/john-travoltage/issues/213#issuecomment-297570162, I would just set focus on the PhET Menu button, since it is the last existing thing that had focus.

**Comment by jessegreenberg** [comment]
@terracoda the above commits send focus back to the PhET menu button when the About dialog and Updates dialog are closed.

Regarding styling of the focus highlight in the PhET menu, I am afraid we will have to wait until the next release to include that feature. :(

#235 should have the "VO konks out" problems fixed.

I think we are ready to close this issue.

**Comment by terracoda** [comment]
Again, great work @jessegreenberg !
